### PR TITLE
Sup cat features

### PR DESCRIPTION
**Pull request by olasaadi**
supporting categorical features

### CODE REVIEW

**Comment by olasaadi** [comment]
I updated the code after I made a pull request, please review the new one

File: apt/minimization/minimizer.py [hunk 1]
```diff
@@ -70,11 +74,14 @@ class GeneralizeToRepresentative(BaseEstimator, MetaEstimatorMixin, TransformerM
     """
 
     def __init__(self, estimator=None, target_accuracy=0.998, features=None,
-                 cells=None):
+                 cells=None, categorical_features=None):
```
Comment (abigailgold) [inline apt/minimization/minimizer.py:77]: Please add a description of the categorical_features param to the docstring of the class

File: apt/minimization/minimizer.py [hunk 2]
```diff
@@ -140,7 +147,7 @@ def fit_transform(self, X=None, y=None):
         self.fit(X, y)
         return self.transform(X)
 
-    def fit(self, X=None, y=None):
+    def fit(self, X: Union[np.ndarray, pd.DataFrame] = None, y=None):
```
Comment (abigailgold) [inline apt/minimization/minimizer.py:150]: Please add this Union typing to the y param as well, and to the fit_transform method too.

File: apt/minimization/minimizer.py [hunk 3]
```diff
@@ -161,7 +168,7 @@ def fit(self, X=None, y=None):
         # take into account that estimator, X, y, cells, features may be None
 
         if X is not None and y is not None:
-            X, y = check_X_y(X, y, accept_sparse=True)
+            # X, y = check_X_y(X, y, accept_sparse=True)
```
Comment (abigailgold) [inline apt/minimization/minimizer.py:171]: Is there somewhere we can move this check to (for example within an if that check it's a numpy array)?

File: apt/minimization/minimizer.py [hunk 4]
```diff
@@ -280,7 +324,7 @@ def transform(self, X):
         check_is_fitted(self, ['cells', 'features'], msg=msg)
 
         # Input validation
-        X = check_array(X, accept_sparse=True)
+        # X = check_array(X, accept_sparse=True)
```
Comment (abigailgold) [inline apt/minimization/minimizer.py:327]: Same comment as before

File: apt/minimization/minimizer.py [hunk 5]
```diff
@@ -315,30 +359,58 @@ def transform(self, X):
...
         for f in self._features:
             if f in cell['ranges']:
                 if not self._cell_contains_numeric(f, cell['ranges'][f], x):
                     return False
+            if f in cell['categories']:
+                if not self._cell_contains_categorical(f, cell['categories'][f], x):
+                    return False
             else:
-                #TODO: exception - feature not defined
+                # TODO: exception - feature not defined
```
Comment (abigailgold) [inline apt/minimization/minimizer.py:387]: Could you please take care of the TODO by adding an appropriate exception?

File: apt/minimization/minimizer.py [hunk 6]
```diff
@@ -315,30 +359,58 @@ def transform(self, X):
...
+                X[feature] = pd.Categorical(X.loc[:, feature], categories=values, ordered=False)
+                ohe = pd.get_dummies(X[feature], prefix=feature)
+                for oneHotVectorFeature in ohe.columns:
+                    self.oneHotVectorFeaturesToFeatures[oneHotVectorFeature] = feature
+                X = pd.concat([X, ohe], axis=1)
+                features_to_remove.append(feature)
+            except KeyError:
+                print('feature not found: ' + feature)
+        self.categorical_data = X.drop(features_to_remove, axis=1)
+        # print(self.data.columns)
```
Comment (abigailgold) [inline apt/minimization/minimizer.py:412]: Please delete all commented out code and prints

File: apt/minimization/minimizer.py [hunk 7]
```diff
@@ -397,21 +478,38 @@ def _calculate_cell_label(self, node):
...
+                    else:
+                        new_cell['ranges'][feature] = {'start': None, 'end': None}
             cells.append(new_cell)
             self.cells_by_id_[new_cell['id']] = new_cell
         self.cells_ = cells
 
     def _calculate_level_cells(self, level):
         if level < 0 or level > self.dt_.get_depth():
-            #TODO: exception 'Illegal level %d' % level
+            # TODO: exception 'Illegal level %d' % level
```
Comment (abigailgold) [inline apt/minimization/minimizer.py:512]: Could you please take care of the TODO by adding an appropriate exception?

File: apt/minimization/minimizer.py [hunk 8]
```diff
@@ -474,45 +572,55 @@ def _get_nodes_level(self, level):
...
             # get all rows in cell
-            indexes = [i for i, x in enumerate(samples_node_ids) if x == cell['id']]
-            sample_rows = samples_df.iloc[indexes]
+            indexes = [i for i, x in enumerate(nodeIds) if x == cell['id']]
+            original_rows = originalTrainFeatures.iloc[indexes]
+            sample_rows = prepared_data.iloc[indexes]
             sample_labels = labels_df.iloc[indexes]['label'].values.tolist()
             # get rows with matching label
             indexes = [i for i, label in enumerate(sample_labels) if label == cell['label']]
+            # match_samples = sample_rows.loc[sample_rows[labelFeature] == cell['label']].drop([labelFeature], axis=1)
```
Comment (abigailgold) [inline apt/minimization/minimizer.py:589]: Please delete all commented out code and prints

File: apt/minimization/minimizer.py [hunk 9]
```diff
@@ -474,45 +572,55 @@ def _get_nodes_level(self, level):
...
             match_samples = sample_rows.iloc[indexes]
+            match_rows = original_rows.iloc[indexes]
             # find the "middle" of the cluster
             array = match_samples.values
+            # Only works with numpy 1.9.0 and higher!!!
+            # median = np.percentile(array, 50, interpolation='nearest', axis=0)
```
Comment (abigailgold) [inline apt/minimization/minimizer.py:595]: Please delete all commented out code and prints

File: apt/minimization/minimizer.py [hunk 10]
```diff
@@ -474,45 +572,55 @@ def _get_nodes_level(self, level):
...
             median = np.median(array, axis=0)
-            # find the record closest to the median
             i = 0
             min = len(array)
+            # print('array length %d' % min)
```
Comment (abigailgold) [inline apt/minimization/minimizer.py:599]: Please delete all commented out code and prints

File: apt/minimization/minimizer.py [hunk 11]
```diff
@@ -474,45 +572,55 @@ def _get_nodes_level(self, level):
...
             min_dist = float("inf")
             for row in array:
                 dist = distance.euclidean(row, median)
                 if dist < min_dist:
                     min_dist = dist
                     min = i
                 i = i + 1
-            row = match_samples.iloc[min]
-            # use its values as the representative
+            # print('min = %d' % min)
```
Comment (abigailgold) [inline apt/minimization/minimizer.py:607]: Please delete all commented out code and prints

File: apt/minimization/minimizer.py [hunk 12]
```diff
@@ -594,15 +714,40 @@ def _get_feature_to_remove(self, samples, nodes, labels, feature_data, current_a
                     range_min = feature_ncp
                     remove_feature = feature
 
+        for feature in categories.keys():
+            # print('trying feature: %s' % feature)
```
Comment (abigailgold) [inline apt/minimization/minimizer.py:718]: Please delete all commented out code and prints

File: apt/minimization/minimizer.py [hunk 13]
```diff
@@ -612,22 +757,44 @@ def _find_range_count(self, samples, ranges):
                 range_counts[r].append(samples_df.shape[0])
             else:
                 for value in ranges[r]:
-                    range_counts[r].append(len(samples_df.loc[samples_df[r] <= value]))
+                    counter = [item for item in samples_df[r] if int(item) <= value]
+                    range_counts[r].append(len(counter))
+                    # range_counts[r].append(len(samples_df.loc[float(samples_df[r]) <= last_value]))
```
Comment (abigailgold) [inline apt/minimization/minimizer.py:762]: Please delete all commented out code and prints

File: apt/minimization/minimizer.py [hunk 14]
```diff
@@ -612,22 +757,44 @@ def _find_range_count(self, samples, ranges):
...
                     last_value = value
-                range_counts[r].append(len(samples_df.loc[samples_df[r] > last_value]))
+                # range_counts[r].append(len(samples_df.loc[float(samples_df[r]) > last_value]))
```
Comment (abigailgold) [inline apt/minimization/minimizer.py:764]: Please delete all commented out code and prints

File: apt/minimization/minimizer.py (hunk 4 above)
Comment (abigailgold) [inline apt/minimization/minimizer.py:327]: Here you also need to check if input is numpy and transform it to pandas

File: apt/minimization/minimizer.py [hunk 15]
```diff
@@ -258,7 +311,7 @@ def fit(self, X=None, y=None):
         # Return the transformer
         return self
 
-    def transform(self, X):
+    def transform(self, X: Union[np.ndarray, pd.DataFrame]):
```
Comment (abigailgold) [inline apt/minimization/minimizer.py:314]: Please update the description of the return value to indicate that it is numpy or pandas according to the input type.

File: apt/minimization/minimizer.py [hunk 16]
```diff
@@ -121,7 +133,7 @@ def set_params(self, **params):
     def generalizations(self):
         return self.generalizations_
 
-    def fit_transform(self, X=None, y=None):
+    def fit_transform(self, X: Union[np.ndarray, pd.DataFrame] = None, y: Union[np.ndarray, pd.DataFrame] = None):
```
Comment (abigailgold) [inline apt/minimization/minimizer.py:136]: Please update the description of the return value to indicate that it is a transformed numpy or pandas array (same as the transform method), and not self.

File: apt/minimization/minimizer.py (hunk 5 above)
Comment (abigailgold) [reply apt/minimization/minimizer.py]: Let's make the string a little more self explanatory, something like "feature " + f + " not defined in cell " + cell['id']

File: tests/test_minimizer.py [hunk 17]
```diff
@@ -1,33 +1,42 @@
...
              {"id": 4, "ranges": {"age": {"start": 39, "end": None}, "height": {"start": 171, "end": None}}, "label": 1,
-              "representative": {"age": 45, "height": 176}}
+              'categories': {}, "representative": {"age": 45, "height": 176}}
              ]
     features = ['age', 'height']
     X = np.array([[23, 165],
-                     [45, 158],
-                     [18, 190]])
+                  [45, 158],
+                  [18, 190]])
```
Comment (abigailgold) [inline tests/test_minimizer.py:37]: Please remove prints from tests. We should replace with asserts where possible.

File: apt/minimization/minimizer.py [hunk 18]
```diff
@@ -227,27 +280,27 @@ def fit(self, X=None, y=None):
                 while accuracy > self.target_accuracy:
                     nodes = self._get_nodes_level(level)
                     self._calculate_level_cells(level)
```
Comment (abigailgold) [inline apt/minimization/minimizer.py:282]: This code either needs to catch the exception thrown by _calculate_level_cells, or check for itself whether level > self.dt_.get_depth() (this is actually another stop condition of the loop).

File: tests/test_minimizer.py (hunk 17 above)
Comment (olasaadi) [reply tests/test_minimizer.py:37]: what asserts can I do on minimizer

File: apt/minimization/minimizer.py (hunk 18 above)
Comment (olasaadi) [reply apt/minimization/minimizer.py]: do you want it to continue running and just break the loop or to finish the run and raise error?

Comment (abigailgold) [reply apt/minimization/minimizer.py]: Break the loop

File: tests/test_minimizer.py (hunk 17 above)
Comment (abigailgold) [reply tests/test_minimizer.py:37]: The whole process is supposed to be deterministic, so you can copy the resulting generalizations and compare to them, for example.

File: apt/minimization/minimizer.py [hunk 19]
```diff
@@ -315,30 +392,57 @@ def transform(self, X):
...
+                values = list(all_values.unique())
+                self.categorical_values[feature] = values
+                X[feature] = pd.Categorical(X.loc[:, feature], categories=values, ordered=False)
+                ohe = pd.get_dummies(X[feature], prefix=feature)
+                for oneHotVectorFeature in ohe.columns:
+                    self.oneHotVectorFeaturesToFeatures[oneHotVectorFeature] = feature
+                X = pd.concat([X, ohe], axis=1)
+                features_to_remove.append(feature)
+            except KeyError:
+                print('feature not found: ' + feature)
```
Comment (abigailgold) [inline apt/minimization/minimizer.py:443]: Let's also please make this print more clear, like what feature was not found and where.

File: tests/test_minimizer.py (hunk 17 above)
Comment (abigailgold) [reply tests/test_minimizer.py:37]: This is great. Can you also add at least one assert for the transformed data (for example that untouched features are not changed, or that features that receive range [] all have the same value)?

Comment (abigailgold) [reply tests/test_minimizer.py:37]: Sorry to keep bugging you about this. The asserts you added are great. I just want one more to test that the data actually is generalized (when it's supposed to). Maybe the easiest way to do this is looking at the ncp scores (make sure that the transformed data has ncp>0). Or, we can compare the columns themselves (same as you did for the non-transformed features, only this time check **not** equals). Thanks!

**Pull request merged**